explorer/server.go

```diff
@@ -0,0 +1,218 @@
+package explorer
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"strconv"
+
+	"github.com/gomodule/redigo/redis"
+	"github.com/gorilla/mux"
+	"github.com/pkg/errors"
+	"github.com/rs/zerolog/log"
+	"github.com/threefoldtech/zos/pkg/rmb"
+)
+
+func (a *App) runServer() {
+	log.Info().Str("Server started ... listening on", string(a.explorer)).Msg("")
+
+}
+
+func (a *App) listFarms(w http.ResponseWriter, r *http.Request) {
+	body, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+		w.Write([]byte("500 - Something bad happened!"))
+		return
+	}
+	log.Debug().Str("request_body", string(body)).Msg("request from external agent")
+	queryString := `
+	{
+		farms {
+			name
+			farmId
+			twinId
+			version
+			farmId
+			pricingPolicyId
+		}
+		publicIps{
+			id
+			ip
+			farmId
+			contractId
+			gateway
+			
+		}
+	}
+	`
+	farmsData := query(queryString)
+	if err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+		w.Write([]byte("500 - Something bad happened!"))
+	}
+	jsonBytes := []byte(farmsData)
+
+	w.Write(jsonBytes)
+}
+
+func (a *App) listNodes(w http.ResponseWriter, r *http.Request) {
+
+	body, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+		w.Write([]byte("500 - Something bad happened!"))
+		return
+	}
+	log.Debug().Str("request_body", string(body)).Msg("request from external agent")
+
+	farmID := r.URL.Query().Get("farm_id")
+	isSpecificFarm := ""
+	if farmID != "" {
+		isSpecificFarm = fmt.Sprintf(",where:{farmId_eq:%s}", farmID)
+	} else {
+		isSpecificFarm = ""
+	}
+	page := r.URL.Query().Get("page")
```

Comment:
Usually (and clean coding) is that optional params like page number of max number of results per page, should be handled by a middleware.
A middleware will basically intercept the request, parse those argument return error if they are invalid, or set defaults if they are missing. then set those on the context of the request object.

Your handler then can get those values (typed) directly from the request context and use those values safely in code. this will allow u to use this middleware with any endpoint that can use pagination instead of having to redo this every time in all handlers